test/Microsoft.AspNet.Razor.Runtime.Precompilation.Test/PrecompilationTagHelperDescriptorFactoryTest.cs

```diff
@@ -0,0 +1,197 @@
+        }
+
+        private static string SanitizeGenericTypeName(string typeName)
+        {
+            return Regex.Replace(typeName, @"\[(?<typename>[^,]+?),[^\]]+?\]", "[${typename}]");
+        }
+
+        private static string ReadManifestResource(string path)
+        {
+            string content;
+            using (var contentStream = ExecutingAssembly.GetManifestResourceStream(path))
+            {
+                using (var reader = new StreamReader(contentStream))
+                {
+                    content = reader.ReadToEnd();
+                }
+            }
+
+            return content;
+        }
+
+        private Compilation GetCompilation(IEnumerable<string> resourceFiles)
+        {
+            var assemblyVersion = TagHelperDescriptorFactoryTestAssembly.Version;
+
+            var syntaxTrees = new List<SyntaxTree>
+            {
+                CSharpSyntaxTree.ParseText(
+                    $"[assembly: {typeof(AssemblyVersionAttribute).FullName}(\"{assemblyVersion}\")]")
+            };
+
+            foreach (var resourceFile in resourceFiles)
+            {
+                var resourceContent = ReadManifestResource(resourceFile);
+                syntaxTrees.Add(CSharpSyntaxTree.ParseText(resourceContent));
+            }
+
+            var libraryExporter = CallContextServiceLocator.Locator.ServiceProvider.GetService<ILibraryExporter>();
+            var applicationName = ExecutingAssembly.GetName().Name;
+            var libraryExport = libraryExporter.GetExport(applicationName);
+
+            var references = new List<MetadataReference>();
+            var roslynReference = libraryExport.MetadataReferences[0] as IRoslynMetadataReference;
+            var compilationReference = roslynReference?.MetadataReference as CompilationReference;
+            if (compilationReference != null)
+            {
+                references.AddRange(compilationReference.Compilation.References);
+                references.Add(roslynReference.MetadataReference);
+            }
+
+            var export = libraryExporter.GetAllExports(applicationName);
+            foreach (var metadataReference in export.MetadataReferences)
+            {
+                // Taken from https://github.com/aspnet/KRuntime/blob/757ba9bfdf80bd6277e715d6375969a7f44370ee/src/...
+                // Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs#L164
+                // We don't want to take a dependency on the Roslyn bit directly since it pulls in more dependencies
+                // than the view engine needs (Microsoft.Framework.Runtime) for example
+                references.Add(ConvertMetadataReference(metadataReference));
+            }
+
+            return CSharpCompilation.Create(
+                GeneratedAssemblyName,
+                syntaxTrees,
+                references);
+        }
+
+        private MetadataReference ConvertMetadataReference(IMetadataReference metadataReference)
```

Comment:
Will see if we can make this a feature of DNX so it doesn't get replicated all over the place.